File: algorithms/1763.longest-nice-substring/solution.rs
use std::collections::HashSet;

impl Solution {
    pub fn longest_nice_substring(s: String) -> String {
        if s.len() < 2 {
            return "".to_string();
        }

        let mut uniqs: HashSet<char> = HashSet::new();
        for c in s.chars() {
            uniqs.insert(c);
        }

        for (i, c) in s.chars().into_iter().enumerate() {
            if uniqs.contains(&c.to_ascii_lowercase()) && uniqs.contains(&c.to_ascii_uppercase()) {
                continue;
            }

            let left = Self::longest_nice_substring(String::from(&s[0..i]));
            let right = Self::longest_nice_substring(String::from(&s[i + 1..]));

            if left.len() >= right.len() {
                return left;
            } else {
                return right;
            }
        }

        s
    }
}

File: algorithms/1763.longest-nice-substring/solution.rb
# @param {String} s
# @return {String}
def longest_nice_substring(s)
  return '' if s.length < 2

  uniqs = s.chars.sort.uniq
  # or
  # uniqs = Set[*s.chars]

  s.each_char.with_index do |c, i|
    next if uniqs.include?(c.upcase) && uniqs.include?(c.downcase)

    left = longest_nice_substring(s[0...i])
    right = longest_nice_substring(s[i + 1..])

    if left.length >= right.length
      return left
    else
      return right
    end
  end

  s
end
